### ISSUE TITLE
Tray refactoring

### ISSUE THREAD

**Issue by NBonaparte** [jaagr/polybar#913]
As is quite obvious by now, the XEMBED tray has presented quite a few issues. By now, it is at best a legacy protocol which has been superseded by [StatusNotifierItem](https://www.freedesktop.org/wiki/Specifications/StatusNotifierItem/) and/or AppIndicator. This would require DBus support, but since we're using Cairo we should be able to leverage GDBus which is part of GLib, a dependency of Cairo.

I've done a little work unrelated to Polybar with this (https://github.com/NBonaparte/sni-tray and https://github.com/NBonaparte/gdbus-sni-watcher), and it seems it might be possible to replace XEMBED altogether. However, this would be a fairly large overhaul with the possibility of breaking some things. Here are some of the main issues:
* Some older applications do not support the protocol. In this case, [`xembed-sni-proxy`](https://github.com/KDE/plasma-workspace/tree/master/xembed-sni-proxy) or something similar should be used to create a dummy XEMBED tray which turns each item into a StatusNotifierItem.
  - Strangely enough, the default `nm-applet` in the Arch linux repos isn't compiled with AppIndicator support. You would need to use `network-manager-applet-indicator` in the AUR and run with `nm-applet --indicator`.

* Menus are supposed to be implemented by the manager itself, rather than the clients (using dbusmenu). If we were using a toolkit like GTK or QT it would be easy, but since we're not, we would have to implement it ourselves.

Rendering the icons should be much less buggy, as with SNI the app simply sends the icon without the background, and the manager (host) would draw it. We could also allow for custom icons to replace the default ones.
Icons also wouldn't move to random places, because we would be the ones handling it.
Scaling should probably be simpler as well.

**Comment by veggiedefender** [comment]
This sounds great. The tray sucks on hidpi displays right now (I'm having some problems with blurry icons and so is #794) and if rewriting this stuff would fix all of that, then I'm all for it.

**Comment by NBonaparte** [comment]
I will probably work on merging #426 first because it implements icon drawing within modules, which would make the new tray much easier to render.

**Comment by IvanMalison** [comment]
@NBonaparte Sorry to ask about this here, but you are the only person I've been able to find who seems to have some clarity on this topic.

Are AppIndicator and SNI the same thing? Is there a formal documentation of their (respective?) protocols (that is more technical that the one you linked) that could be used as a reference for implementing this?

**Comment by IvanMalison** [comment]
I suspect that you guys would be reluctant to rely on anything written in haskell, but I just thought that I would let you know that I've created a standalone watcher executable in:

https://github.com/IvanMalison/status-notifier-item-tray

The watcher exists independently of the host (polybar in your case) in SNI (and for good reason) but I imagine that this might intimidate a certain class of users (why do I have to start an extra process just to run a tray? I don't need to do that with xembed...). I'm curious to hear what your thoughts on this are.

**Comment by NBonaparte** [comment]
I also made a watcher a while back in C: https://github.com/NBonaparte/gdbus-sni-watcher

Although I haven't done much work on this feature recently, I believe I intended to spawn the watcher as a separate process, after checking for its existence. That way multiple trays can run with one watcher.

**Comment by crocket** [comment]
@IvanMalison I think it makes sense to not have an extra process even if SNI is used.

**Comment by IvanMalison** [comment]
having an extra process is much better IMO, bc it makes the whole system more robust in that registration and "hosting" are decoupled. Some SNIs are not particularly good about listening for new status notifier watchers, so if you restart the process for watching, it can take a few moments for all the SNIs to register.

It's pretty easy to make it user configureable though, so that people can just the configuration that they prefer. My plan is to add an option to https://github.com/IvanMalison/gtk-sni-tray that also starts a watcher within the same process

**Comment by NBonaparte** [comment]
I've also noticed that if you kill and respawn a watcher, electron apps will segfault, which implies that they expect a watcher to be running all the time. This can be somewhat troublesome if you have it as part of the same process, as reloading the bar/tray (if monitors are reconfigured, for example) would cause this to occur, unless changing positions/etc. can be done on the fly.

**Comment by crocket** [comment]
I prefer serverless auto configuration, but it is what it is. SNI is not serverless.

**Comment by IvanMalison** [comment]
> SNI is not serverless

SNI isn't really an option/not option. It's the only future proof game in town.

**Comment by crocket** [comment]
@IvanMalison As far as I know, you work on taffybar. Are you moving away from taffybar to polybar?

**Comment by IvanMalison** [comment]
No, I'm just following this issue because I'm work on adding SNI to taffybar right now, and I want to see whats going on in similar projects with that. At this point, I'm actually quite far along; The bar and menus actually work and everything, but there are a few kinks to iron out.

**Comment by IvanMalison** [comment]
Taffybar [SNI support ](https://github.com/travitch/taffybar/blob/master/src/System/Taffybar/Widgets/SNITray.hs) is finished, and I just thought I'd let you guys know about some of the things that came up during development just in case it can help save you some time (and I generally think reciprocal knowledge sharing can be helpful).

- Unfortunately, starting the watcher process automatically when the host is started does not seem to be a great approach, because most SNI Items don't actually watch for new watchers. This means that live SNIs wont necessarily register with your watcher at startup if they started before your watcher, thus it is generally really important to have the watcher start early and be a super stable, reliable process.

I'm not sure what peoples usage is like, but I often find myself restarting taffybar to tweak things, so it is generally best to have the watcher isolated in a long running and very stable process.

- Different items seem to implement different specs. chrome, for example, does something super weird where it basically passes the file path and the item name under two different parameter names. I ended up using gtk to load icons for me, which works for most cases, but its definitely a bit tricky. What are you guys going to do (since I imagine you don't plan on adding gtk as a dep)

**Comment by NBonaparte** [comment]
Thanks for sharing. I've been busy recently so I have not had a lot of time to work on this. 

So, would the best method be to make an independent binary which the user can put in their WM/DE/etc. startup script?

I've also noticed a fair amount of variations on the spec; I once read through the code of appindicator (or something similar, can't remember) and found a lot of different cases which the host has to sort through. We could check for each noncompliant item, but that is somewhat tedious. I came up with a somewhat hacky method a while back for deducing the GTK icon theme and path, but it's probably less than ideal.
GTK support is becoming increasingly attractive, as we need to support dbusmenu, and implementing a menuing system is probably overkill.

**Comment by IvanMalison** [comment]
Yeah I ended up using libdbusmenugtk. dbusmenu is a pretty complicated interface... I don't think I'd want to implement that whole thing myself.

The thing about adding gtk for you guys is that im not sure how well it will interact with everything else. I'm sure someone has tried it before though.
